### PR TITLE
Add bucket sns notification support to attach-encrypt

### LINKED ISSUE #297: attach-encrypt should support bucket sns notification

### PR DESCRIPTION
**Pull request by whit537**
Blorg, sorry. Started over and wasn't careful to preserve the branch. This is a #910 redux. Closes #297.

### CODE REVIEW

**Comment by whit537** [comment]
I've pushed an implementation w/ test for the new topic case. Next is the existing topic case ...

**Comment by whit537** [comment]
@kapilt Can you give me some insight on what the [`S3_AUGMENT_TABLE`](https://github.com/capitalone/cloud-custodian/blob/957c450011f49f2bff02f6e7f1cae26de3f4a082/c7n/resources/s3.py#L111-L123) is doing? I believe it may be implicated in an error I'm [hung up on](https://github.com/whit537/cloud-custodian/issues/1#issuecomment-279991601).

**Comment by whit537** [comment]
This bug [affects](https://github.com/whit537/cloud-custodian/issues/1#issuecomment-280318365) an existing test, not just the new one I made. Debugging ...

**Comment by kapilt** [comment]
commented on the referenced issue link.

**Comment by whit537** [comment]
Rebased and ready for review, @kapilt.

File: docs/source/usecases/s3encryption.rst [hunk 1]
```diff
@@ -14,6 +17,59 @@ remediate them such that they are encrypted.
...
+encrypted.
+
+.. code-block:: yaml
+
+   policies:
+     - name: s3-attach-encryption
+       resource: s3
+       actions:
+         - type: attach-encrypt
+           role: arn:aws:iam::123456789012:role/my-role
```
Comment (kapilt) [inline docs/source/usecases/s3encryption.rst]: the role aspect seems to be confusing people wrt to the same doc item used in config-rule. note the role is optional if folks are using --assume-role on the command-line. worth noting.

File: c7n/resources/s3.py [hunk 2]
```diff
@@ -707,10 +711,49 @@ def process(self, buckets):
...
+                        'Action': 'SNS:Publish',
+                        'Effect': 'Allow',
+                        'Resource': topic_arn,
+                        'Principal': {'Service': 's3.amazonaws.com'}}]}
+                sns.set_topic_attributes(
+                    TopicArn=topic_arn,
+                    AttributeName='Policy',
+                    AttributeValue=json.dumps(policy))
+                session.resource('s3').BucketNotification(bname).put(
+                    NotificationConfiguration={'TopicConfigurations': [{
```
Comment (kapilt) [inline c7n/resources/s3.py]: just realized this is an issue as well with the extant lambda function notification setup, in that we're effectively blanking extant notify config :-(

File: c7n/resources/s3.py [hunk 3]
```diff
@@ -634,7 +634,9 @@ class AttachLambdaEncrypt(BucketActionBase):
                   - attach-encrypt
     """
     schema = type_schema(
-        'attach-encrypt', role={'type': 'string'})
+        'attach-encrypt',
+        role={'type': 'string'},
+        via={'type': 'string', 'enum': ['bucket-notification', 'sns']})
```
Comment (kapilt) [inline c7n/resources/s3.py]: their both bucket-notifications, just one is lambda/function

File: tests/test_s3.py [hunk 4]
```diff
@@ -619,6 +620,24 @@ def test_attach_encrypt_requires_role(self):
              'resource': 's3',
              'actions': [{'type': 'attach-encrypt'}]})
 
+    @skip_if_not_validating
```
Comment (kapilt) [inline tests/test_s3.py]: re skip_if_not.. seems fine, note load_policy also has options to validate

**Comment by kapilt** [comment]
let's pull up on gitter to discuss this branch this week.

**Comment by whit537** [comment]
Per call, I'll spend a little more time with this to see if it's close or still far off.

File: tests/test_s3.py (hunk 4 above)
Comment (whit537) [inline tests/test_s3.py]: The reason to skip is asymmetry between Travis and Drone. We [set](https://github.com/capitalone/cloud-custodian/blob/afe83481a95422df306fc7ef71a355b2069029fd/.travis.yml#L6) `C7N_VALIDATE=yes` at the former but [not](https://github.com/capitalone/cloud-custodian/blob/afe83481a95422df306fc7ef71a355b2069029fd/.drone.yml#L5-L8) at the latter. See https://github.com/whit537/cloud-custodian/issues/1#issuecomment-279134438 ff. Alternative would be to standardize test environments one way or another.

File: c7n/resources/s3.py (hunk 3 above)
Comment (whit537) [inline c7n/resources/s3.py]: Okay. How about `cloud-function` and `topic` per the XML examples in [this doc](http://docs.aws.amazon.com/AmazonS3/latest/dev/NotificationHowTo.html)?

**Comment by whit537** [comment]
Rebased, was afe83481a95422df306fc7ef71a355b2069029fd.

File: docs/source/usecases/s3encryption.rst (hunk 1 above)
Comment (whit537) [inline docs/source/usecases/s3encryption.rst]: Addressed in 2422883df9d63ed2a831b8bad2f499e43873fefc.

File: c7n/resources/s3.py (hunk 3 above)
Comment (whit537) [inline c7n/resources/s3.py]: Done in 2422883df9d63ed2a831b8bad2f499e43873fefc.

File: c7n/ufuncs/s3crypt.py [hunk 5]
```diff
@@ -41,28 +41,36 @@ def init():
...
-            # Ensure we know which key caused an issue
-            print("error %s:%s code:%s" % (
-                bucket, key['Key'], e.response['Error']))
-            raise
-        if not result:
-            return
-        print("remediated %s:%s" % (bucket, key['Key']))
+    for outer in event.get('Records', []):
+        print outer
+        if outer.get('EventSource', '') == 'aws:sns':
```
Comment (whit537) [inline c7n/ufuncs/s3crypt.py]: Stray print shot in 2422883df9d63ed2a831b8bad2f499e43873fefc.

Comment (whit537) [inline c7n/ufuncs/s3crypt.py]: `process_key_event` refactored in e7bb5230cbfabd6fc41a0b85448c29e5b8d2a52b.

**Comment by whit537** [comment]
Rebased on latest `master` and reduced to two commits, with d7aa2ee3939e62e2b3d11434b5a6685ed6051ead being the interesting one. Previous head was f35467eb7c1d8ac8a108b6cd78d4aabc5f177294.

**Comment by whit537** [comment]
This is ready for review, @kapilt @ocampocj.

**Comment by whit537** [comment]
Rebased on master, was b2b2ecc0c2aa171a27a1461d3113b61399dd74be.

**Comment by kapilt** [comment]
looks good, thanks. one request, could you slim down the recorded test data files, there's quite a few resulting from the list buckets call as the other buckets in the account are causing a few per bucket calls that can be omitted.

**Comment by whit537** [comment]
Rebased on lastest master (was 35038d17f47be9d64cbfdf926bf64903ac35cade), looking at slimming down fixture ...

**Comment by kapilt** [comment]
fwiw best way to slim down is a manual edit or mock, manual edit, kill extra buckets in listing, and reorder json responses on individual calls (decrements on sequence numbers in file names).

**Comment by whit537** [comment]
How about 74c8b24d82a2cd574f3a91f0e044963f0d88893a, @kapilt? I see 27% fewer files.

screen shot 2017-04-24 at 7 28 41 am
screen shot 2017-04-24 at 7 28 47 am

**Comment by kapilt** [comment]
much better, thanks

File: c7n/mu.py [hunk 6]
```diff
@@ -1058,6 +1058,49 @@ class Done(Exception): pass
...
+        session = local_session(self.session_factory)
+        sns = session.client('sns')
+        s3 = session.client('s3')
+
+        notifies = bucket['Notification']
+        if 'TopicConfigurations' not in notifies:
+            notifies['TopicConfigurations'] = []
+        all_topics = notifies['TopicConfigurations']
+        topic_arns = [t['TopicArn'] for t in all_topics
+                      if 's3:ObjectCreated:*' in t['Events']]
```
Comment (kapilt) [inline c7n/mu.py]: one note, this is a lot of implicit policy on the creation of the topic, at the mu.py level it shouldn't really be creating topics or care about which events folks want to subscribe to.

**Pull request merged**